### ISSUE TITLE
Support Fastify

### ISSUE THREAD

**Issue by SimenB** [remcohaszing/axios-test-instance#2]
We're currently migrating from `express` to `fastify`, would love to keep using this library!

https://www.fastify.io/

**Comment by remcohaszing** [comment]
Thank you for your interest!

I gave it a try, but as you can see in the failing tests of the pull request, starting a fastify instance a second time, even after closing it properly, fails.

Also I’m not too happy about the framework detection mechanism I came up with.

I don’t think I’m going to dive much deeper into support for Fastify, but if someone else manages to fix it, I’ll gladly merge it.

**Comment by SimenB** [comment]
In their docs, they have a `supertest` example which gets passed the `server`. Would that work, somehow?

https://www.fastify.io/docs/v3.0.x/Testing/#example-1

Seems we'd have to do `ready` and `close` ourselves regardless, which somewhat defeats the point... 🙁

**Comment by remcohaszing** [comment]
> Seems we'd have to do ready and close ourselves regardless, which somewhat defeats the point... :slightly_frowning_face: 

That part has already been abstracted away. I want this library to make the test code as simple as possible. :wink: 

---

The issue is that fastify instances can’t be reused. I.e. the following code won’t work if multiple tests are present.

```js
// app.js
import * as Fastify from 'fastify'

const app = Fastify();
app.get('/', async (request, reply) => {
  reply.send({ hello: 'world' });
});
export default app;

// app.test.js
import { request, setTestApp } from 'axios-test-instance'
import app from './app'

beforeEach(() => setTestApp(app))
```

The following examples do work with the current PR. This is the same pattern as is used in their supertest example.

```js
// createApp.js
import * as Fastify from 'fastify'

export default function createApp() {
  const app = Fastify();
  app.get('/', async (request, reply) => {
    reply.send({ hello: 'world' });
  });
  return app;
}

// createApp.test.js
import { request, setTestApp } from 'axios-test-instance'
import createApp from './createApp'

beforeEach(() => setTestApp(createApp()))
```

```js
// createApp.js
import * as Fastify from 'fastify'

export default function createApp() {
  const app = Fastify();
  app.get('/', async (request, reply) => {
    reply.send({ hello: 'world' });
  });
  return app;
}

// createApp.test.js
import { request, setTestApp } from 'axios-test-instance'
import createApp from './createApp'

beforeAll(() => setTestApp(createApp()))
```

Because jest creates new instances for each test file when module isolation isn’t disabled, the following should also just work with jest.

```js
// app.js
import * as Fastify from 'fastify'

const app = Fastify();
app.get('/', async (request, reply) => {
  reply.send({ hello: 'world' });
});
export default app;

// app.test.js
import { request, setTestApp } from 'axios-test-instance'
import app from './app'

beforeAll(() => setTestApp(app))
```

---

If the scenario above meets your requirements, I’ll gladly finalize my changes and release them with a footnote about the quirks of using this library with fastify.

**Comment by SimenB** [comment]
Yeah, that should cover my use cases, thanks! I think documenting that behavior should be enough

**Comment by remcohaszing** [comment]
This has been released as 3.1.0.

**Comment by SimenB** [comment]
Thanks! It works great, except for https://github.com/remcohaszing/axios-test-instance/pull/3/files#r439827863
